minio/operator
# Control the start of event reporting for expiring certificates

### Issue summary

In cases where users install certificates of short duration, kubernetes events are spammed with multiple warnings of certificate expiry. This is undesirable. A means on reducing the number of warnings produced is more desirable.
See https://github.com/minio/operator/issues/1791

In the tenant crd, `.spec.certExpiryAlertThreshold` represents the minimum number of days to expiry before an alert for an expiring certificate is fired.
For example considering `.spec.certExpiryAlertThreshold = 1`, if a given certificate will expire in 7 days then expiration events will only be triggered 1 day before expiry.


### Test summary

1. Build and install the minio operator
2. Apply a tenant with Helm or Kustomize
3. Create expired and expiring certificates
4. Create secrets based on these
5. Apply the secrets as external certificates
6. Experiment with different values of `certExpiryAlertThreshold` e.g.
7. Observe

* if certExpiryAlertThreshold is not present then all certificate expiry events are reported. This is current functionality
* if the whole number of days left for a certificate to expire is less than certExpiryAlertThreshold then the certificate expiry events are reported
* if the whole number of days left for a certificate to expire is greater than or equal to certExpiryAlertThreshold then the certificate expiry events are not reported
* expired certificate events are always reported

Verbose tests here:
https://github.com/allanrogerr/public/wiki/operator%E2%80%901791

Fixes: https://github.com/minio/operator/issues/1791
